{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Agent-based model of boundary-driven grouth\n",
    "To model the expected relationship between genomic divergence and spatial diversity, I used a cell automous model based on the one described in \"Evolution and Phenotypic Selection of Cancer Stem Cells\" (Poleszczuk, Hahnfeldt, Enderling, PLOS Computational Biology, 2015). The simulation is initiated with a single cell in the center of the lattice. At each time point $\\Delta t$, a cell attempts division with rate $\\lambda$ or dies with rate $\\alpha$. Under boundary-driven growth, cells only successfully divide if there is an empty lattice spot in its Moore neighborhood. If multiple neighboring spaces are available then the cell randomly chooses the location for its daughter cell. The result of cell division is two cells, one of which is in the same location as the parent cell and one of which is in a new lattice square.  In the unrestricted mode of growth, if a cell attempts division, its daughter cell will occupy an empty lattice spot in the Moore neighborhood if available, but if not, the cell will still divide and push cells in a random direction to make space. Overlapping cells are pushed in the same direction until a neighboring lattice spot is available, which the pushed cell will occupy. In both simulations, if a cell divides, each daughter cell can gain a mutations with probability $\\mu$. Mutations are then drawn from Jukes-Cantor model of sequence evolution and follow an infinite-sites assumption. The endpoint of the simulations are defined by the cell total number of alive cells. The ground-truth growth rate is assessed at time slices in the simulation and recording the current state of each cell and the probability that the cell has a progeny in the next time step. Effective spatial constraints in the boundary-driven model were controlled by changing cell death rate, where increased cell turnover allows cells to divide more readily. Input parameters are proliferation and death scalars, where $\\lambda = \\Delta t(proliferation - death)/2$ and $\\alpha = \\Delta t(death)/2$.To evaluate the accuracy of parameter estimation, we ran 1000-cell tumor simulations with $proliferation = 1$, $\\mu = 1$ and a range of $death = (0-0.87)$. Simulation will restart if the entire population dies out. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setup\n",
    "import itertools\n",
    "import operator\n",
    "import copy\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import operator\n",
    "from matplotlib import pyplot as plt\n",
    "import matplotlib as mpl\n",
    "\n",
    "from statsmodels.nonparametric.api import KernelReg\n",
    "import pickle\n",
    "import random\n",
    "import matplotlib.cm as cm\n",
    "import os\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define cancer simulation \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#function to check if adjacent cells are empty\n",
    "\n",
    "def adjacent_cell(lattice,i,j):\n",
    "    free_cells = []\n",
    "    change = [1,-1]\n",
    "    for c in change:\n",
    "        checki = lattice[i+c,j]\n",
    "        if checki == None:\n",
    "            free_cells.append([i+c,j])\n",
    "        checkj = lattice[i, j+c]\n",
    "        if checkj == None:\n",
    "            free_cells.append([i,j+c])\n",
    "        for h in change:\n",
    "            checkd = lattice[i+h,j+c]\n",
    "            if checkd == None:\n",
    "                free_cells.append([i+h,j+c])\n",
    "                \n",
    "    return free_cells\n",
    "\n",
    "def cancer_sim(founder_cells = 1,deltat=(1/float(24)),CSC = True,max_cells = 1000,\n",
    "               proliferationCSC = 1, proliferationCC = 2, motility = 15, pmax = 10,\n",
    "               mortality = 0.1, mutation = 0.1, selection = False,ps = 0.05,\n",
    "              free_diffusion = False, JC = False, s = 0.1, driverMutCell = 100):    \n",
    "\n",
    "    #create classes\n",
    "    class Cell:\n",
    "        def __init__(self):\n",
    "            \n",
    "            self.parent_index = 0 #index of cell it was derived from (for df conversion)\n",
    "\n",
    "            self.locx = None #x location in lattice\n",
    "            self.locy = None #y location in lattice\n",
    "            self.birthdate = None\n",
    "            self.deathdate = None\n",
    "            self.pmax = pmax #proliferation potential\n",
    "            self.cellnum = None #unique cell number to identify\n",
    "            self.alpha = mortality #probability of sponaneous death\n",
    "            self.mutation_rate = mutation #probability of mutation per cell division\n",
    "\n",
    "            self.proliferation_rate = proliferationCC #average cell divisions per time unit\n",
    "\n",
    "            self.mutations = [] #store index of mutations\n",
    "            self.sequence = [] #store simulated nucleotide sequence under JC model\n",
    "            self.ps = 0\n",
    "            self.cells_free = 8 #add to keep track of spatial constraint at time of cell birth\n",
    "            self.driver = 0\n",
    "\n",
    "\n",
    "#function to convert results into dictionary, when can be converted to dataframe\n",
    "        def as_dict(self):\n",
    "\n",
    "\n",
    "            return {'index': self.cellnum,\n",
    "                    'parent_index': self.parent_index,\n",
    "                    'birthdate': self.birthdate,\n",
    "                    'locx': self.locx,\n",
    "                    'locy': self.locy,\n",
    "                    'deathdate': self.deathdate,\n",
    "                    'mutations': self.mutations,\n",
    "                    'proliferation_rate': self.proliferation_rate,\n",
    "                    'alpha': self.alpha,\n",
    "                    'mutation_rate': self.mutation_rate, \n",
    "                    'cells_free':self.cells_free, \n",
    "                    'sequence': self.sequence,\n",
    "                    'driver': self.driver}\n",
    "\n",
    "\n",
    "    class StemCell(Cell): #define stem cell class\n",
    "        def __init__(self):\n",
    "            Cell.__init__(self)\n",
    "            self.pmax = float('inf') #infinite proliferation potential\n",
    "            self.alpha = 0 #immortal\n",
    "            self.ps = ps #probability of symmetric division\n",
    "            self.proliferation_rate = proliferationCSC\n",
    "\n",
    "\n",
    "    #create lattice array\n",
    "    N = max_cells # starting demensions of lattice\n",
    "    lattice = np.empty( (N,N), dtype=object)\n",
    "\n",
    "    #parameters\n",
    "    proliferation_rate_CC = proliferationCC #proliferation rate of clonal cell\n",
    "    proliferation_rate_CSC = proliferationCSC #proliferation rate of cancer stem cell\n",
    "    pmax_CC = pmax\n",
    "\n",
    "    alpha_CC = mortality\n",
    "    cur_cellnum = 1\n",
    "    cur_mutnum = 1\n",
    "\n",
    "    mutation_rate = mutation\n",
    "    time = 0\n",
    "    \n",
    "    #initialize list to keep track of cells (all time and alive)\n",
    "    alive_cells = []\n",
    "    cells = []\n",
    "    \n",
    "    nucleotides = [\"A\", \"G\", \"C\", \"T\"]\n",
    "\n",
    "#to place founder cell in center of lattice\n",
    "    i = int(N/2)\n",
    "    modi = [0,0,1,1]\n",
    "    modj = [0,1,0,1]\n",
    "    \n",
    "    if CSC == True:\n",
    "        for x in range(founder_cells):\n",
    "            Fcell = StemCell() #initiate founder cells, stem-cells\n",
    "            Fcell.ps = ps\n",
    "            Fcell.proliferation_rate = proliferation_rate_CSC #average cell divisions per day\n",
    "            Fcell.locx = i+modi[x] #x location in lattice\n",
    "            Fcell.locy = i+modj[x] #y location in lattice\n",
    "            Fcell.birthdate = 0\n",
    "            Fcell.cellnum = cur_cellnum #unique cell number to identify\n",
    "            cur_cellnum += 1\n",
    "            Fcell.mutation_rate = mutation_rate #probability of mutation per cell division\n",
    "\n",
    "\n",
    "            cells.append(Fcell)\n",
    "            alive_cells.append(Fcell)\n",
    "\n",
    "            lattice[Fcell.locx,Fcell.locy] = Fcell\n",
    "    else:\n",
    "        for x in range(founder_cells):\n",
    "            Fcell = StemCell() #initiate founder cells, non-stem cells\n",
    "            Fcell.proliferation_rate = proliferation_rate_CC\n",
    "            Fcell.locx = i+modi[x] #x location in lattice\n",
    "            Fcell.locy = i+modj[x] #y location in lattice\n",
    "            Fcell.alpha = mortality\n",
    "            Fcell.ps = 0\n",
    "            Fcell.birthdate = 0\n",
    "            Fcell.cellnum = cur_cellnum  #unique cell number to identify\n",
    "            cur_cellnum += 1\n",
    "            Fcell.mutation_rate = mutation_rate #probability of mutation per cell division\n",
    "\n",
    "\n",
    "            cells.append(Fcell)\n",
    "            alive_cells.append(Fcell)\n",
    "            lattice[Fcell.locx,Fcell.locy] = Fcell\n",
    "\n",
    "\n",
    "    #time parameters\n",
    "    dt = deltat # time is equilavent to 1/24 of a day or 1 hour\n",
    "   \n",
    "    while len(alive_cells) < max_cells:\n",
    "        if len(alive_cells) < 1:\n",
    "            print('no cells alive')\n",
    "            #start overr simulation\n",
    "            cells = []\n",
    "            alive_cells = []\n",
    "            time = 0\n",
    "            cur_cellnum = 1\n",
    "            cur_mutnum = 1\n",
    "\n",
    "            lattice = np.empty( (N,N), dtype=object)\n",
    "            for x in range(founder_cells):\n",
    "                Fcell = StemCell() #initiate founder cells, non-stem cells\n",
    "                Fcell.proliferation_rate = proliferation_rate_CC\n",
    "                Fcell.locx = i+modi[x] #x location in lattice\n",
    "                Fcell.locy = i+modj[x] #y location in lattice\n",
    "                Fcell.alpha = mortality\n",
    "                Fcell.ps = 0\n",
    "                Fcell.birthdate = 0\n",
    "                Fcell.cellnum = cur_cellnum  #unique cell number to identify\n",
    "                cur_cellnum += 1\n",
    "                Fcell.mutation_rate = mutation_rate #probability of mutation per cell division per time unit\n",
    "\n",
    "\n",
    "                cells.append(Fcell)\n",
    "                alive_cells.append(Fcell)\n",
    "                lattice[Fcell.locx,Fcell.locy] = Fcell\n",
    "            \n",
    "            \n",
    "        time += dt\n",
    "        cell_stack = random.sample(alive_cells,len(alive_cells)) #random order of cells\n",
    "        \n",
    "        time += dt\n",
    "        cell_stack = random.sample(alive_cells,len(alive_cells)) #random order of cells\n",
    "                \n",
    "        for cell in cell_stack:\n",
    "            alive = True\n",
    "            r = random.uniform(0, 1) #decision for prolferation\n",
    "            r2 = random.uniform(0, 1) #decision for cell death\n",
    "            pd = cell.proliferation_rate * dt #probability of proliferation in time dt\n",
    "            \n",
    "            if free_diffusion: #for free diffusion model cell is naive to neighbors in movement\n",
    "                free_cells = [[cell.locx + 1, cell.locy + 1],\n",
    "                             [cell.locx - 1, cell.locy + 1],\n",
    "                             [cell.locx + 1, cell.locy - 1],\n",
    "                             [cell.locx - 1, cell.locy - 1]]\n",
    "            else:\n",
    "                free_cells = adjacent_cell(lattice,cell.locx,cell.locy)\n",
    "                \n",
    "            p_die = cell.alpha * dt #probability of cell death in time dt\n",
    "            \n",
    "            if r < p_die:\n",
    "                alive = False\n",
    "                lattice[cell.locx,cell.locy] = None\n",
    "                alive_cells.remove(cell)\n",
    "\n",
    "                cell.deathdate = time\n",
    "            \n",
    "                \n",
    "            elif r2 < pd: # Does cell attempt to divide? \n",
    "                if len(free_cells) > 0: #is there any space to divide?\n",
    "\n",
    "                    if cell.pmax > 0: #is cell proliferation capacity exhausted?\n",
    "\n",
    "                        cell.pmax -= 1\n",
    "                        r = random.uniform(0,1)\n",
    "                        new_cell = copy.deepcopy(cell)\n",
    "                        new_cell_2 = copy.deepcopy(cell)\n",
    "                        \n",
    "                        if r <= cell.ps: #does cell divide asymmetrically?\n",
    "                            new_cell.pmax = pmax\n",
    "                            new_cell.proliferation_rate = proliferation_rate_CC #average cell divisions per day\n",
    "                            new_cell.alpha = mortality\n",
    "                            new_cell.ps = 0\n",
    "\n",
    "                        newloc = random.choice(free_cells)\n",
    "                        new_cell.locx = newloc[0] #x location in lattice\n",
    "                        new_cell.locy = newloc[1] #y location in lattice\n",
    "                        new_cell_2.locx = cell.locx #x location in lattice\n",
    "                        new_cell_2.locy = cell.locy #y location in lattice\n",
    "                        new_cell.birthdate = time\n",
    "                        new_cell_2.birthdate = time\n",
    "                        new_cell.cellnum = cur_cellnum #unique cell number to identify\n",
    "                        \n",
    "                        #add selection, when driver mutation occurs is dependent on driverMutCell\n",
    "                        if cur_cellnum == driverMutCell and selection:\n",
    "                            \n",
    "                            new_cell.proliferation_rate = (1+s)* new_cell.proliferation_rate\n",
    "                            new_cell.driver = 1\n",
    "                    \n",
    "                        cur_cellnum += 1\n",
    "                        new_cell_2.cellnum = cur_cellnum\n",
    "                        new_cell.parent_index = cell.cellnum\n",
    "                        new_cell_2.parent_index = cell.cellnum\n",
    "                        \n",
    "\n",
    "                        if cur_cellnum == driverMutCell and selection:\n",
    "                            \n",
    "                            new_cell_2.proliferation_rate = (1+s)* new_cell_2.proliferation_rate\n",
    "                            new_cell_2.driver = 1\n",
    "                        \n",
    "                        cur_cellnum += 1\n",
    "                        \n",
    "                        lattice[new_cell.locx,new_cell.locy] = new_cell\n",
    "                        lattice[cell.locx,cell.locy] = new_cell_2\n",
    "                        \n",
    "                        if not free_diffusion:\n",
    "                            new_cell.cells_free = len(adjacent_cell(lattice,new_cell.locx,new_cell.locy))\n",
    "                            new_cell_2.cells_free = len(adjacent_cell(lattice,new_cell_2.locx,new_cell_2.locy))\n",
    "\n",
    "                        r1 = random.uniform(0,1)\n",
    "                        if r1 < cell.mutation_rate: #does cell gain a mutation?\n",
    "                            \n",
    "                            if JC: #if Jukes-Cantor model\n",
    "                                \n",
    "                                original_nucleotide = random.choice([\"A\", \"G\", \"C\", \"T\"])\n",
    "                                \n",
    "                                #put original nucleotide in all sequences\n",
    "                                for cell_0 in cells:\n",
    "                                    cell_0.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                                #cell 2 has not been added yet\n",
    "                                new_cell_2.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                                #replace mutation in this cells sequence\n",
    "                                \n",
    "                                possible_nucs = [\"A\", \"G\", \"C\", \"T\"]\n",
    "                                possible_nucs.remove(original_nucleotide)\n",
    "                                new_nucleotide = random.choice(possible_nucs)\n",
    "                                new_cell.sequence.append(new_nucleotide)\n",
    "                            \n",
    "                            new_mut = cur_mutnum\n",
    "                            cur_mutnum +=1\n",
    "                            new_cell.mutations.append(new_mut)\n",
    "                        \n",
    "                        #does second cell get a mutation\n",
    "                        r2 = random.uniform(0,1)\n",
    "                        if r2 < cell.mutation_rate: #does cell gain a mutation?\n",
    "                            \n",
    "                            if JC: #if Jukes-Cantor model\n",
    "                                \n",
    "                                original_nucleotide = random.choice([\"A\", \"G\", \"C\", \"T\"])\n",
    "                                \n",
    "                                #put original nucleotide in all sequences\n",
    "                                for cell_0 in cells:\n",
    "                                    cell_0.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                                new_cell.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                                possible_nucs = [\"A\", \"G\", \"C\", \"T\"]\n",
    "                                possible_nucs.remove(original_nucleotide)\n",
    "                                new_nucleotide = random.choice(possible_nucs)\n",
    "                                new_cell_2.sequence.append(new_nucleotide)\n",
    "\n",
    "                            new_mut = cur_mutnum\n",
    "                            cur_mutnum +=1\n",
    "                            new_cell_2.mutations.append(new_mut)\n",
    "\n",
    "\n",
    "                        alive_cells.append(new_cell)\n",
    "                        alive_cells.append(new_cell_2)\n",
    "                        cells.append(new_cell)\n",
    "                        cells.append(new_cell_2)\n",
    "                        alive_cells.remove(cell)\n",
    "\n",
    "                        cell.deathdate = time\n",
    "\n",
    "\n",
    "    for cell in alive_cells:\n",
    "\n",
    "        cell.deathdate = time + dt #need to add one time step to differentiate cells that died in last gen\n",
    "\n",
    "\n",
    "    return cells, alive_cells, lattice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Simulation with pushing\n",
    "##Same as above except simulation creates new cell when pushed\n",
    "##Also saves locations through time since they can change during a cells lifespan\n",
    "\n",
    "def pushing_cancer_sim2(founder_cells = 1,deltat=(1/float(24)),CSC = True,max_cells = 1000,\n",
    "               proliferationCSC = 1, proliferationCC = 2, motility = 15, pmax = 10,\n",
    "               mortality = 0.1, mutation = 0.1, selection = False, JC = True):    \n",
    "\n",
    "    #create classes\n",
    "    class Cell:\n",
    "        def __init__(self):\n",
    "            \n",
    "            self.parent_index = 0 #index of cell it was derived from (for df conversion)\n",
    "\n",
    "            self.locx = None #x location in lattice\n",
    "            self.locy = None #y location in lattice\n",
    "            self.birthdate = None\n",
    "            self.deathdate = None\n",
    "            self.pmax = pmax #proliferation potential\n",
    "            self.cellnum = None #unique cell number to identify\n",
    "            self.alpha = mortality #probability of sponaneous death\n",
    "            self.mutation_rate = mutation #probability of mutation per cell division\n",
    "\n",
    "            self.proliferation_rate = proliferationCC #average cell divisions per time unit\n",
    "\n",
    "            self.mutations = [] #store index of mutations\n",
    "            self.sequence = [] #store simulated nucleotide sequence under JC model\n",
    "            #self.ps = 0\n",
    "            self.cells_free = 8 #add to keep track of spatial constraint at time of cell birth\n",
    "\n",
    "\n",
    "#function to convert results into dictionary, when can be converted to dataframe\n",
    "        def as_dict(self):\n",
    "\n",
    "\n",
    "            return {'index': self.cellnum,\n",
    "                    'parent_index': self.parent_index,\n",
    "                    'birthdate': self.birthdate,\n",
    "                    'locx': self.locx,\n",
    "                    'locy': self.locy,\n",
    "                    'deathdate': self.deathdate,\n",
    "                    'mutations': self.mutations,\n",
    "                    'proliferation_rate': self.proliferation_rate,\n",
    "                    'alpha': self.alpha,\n",
    "                    'mutation_rate': self.mutation_rate, \n",
    "                    'cells_free':self.cells_free, \n",
    "                    'sequence': self.sequence}\n",
    "        \n",
    "        def get_loc_dict(self):\n",
    "\n",
    "\n",
    "            return {'index': self.cellnum,\n",
    "                    'locx': self.locx,\n",
    "                    'locy': self.locy}\n",
    "\n",
    "\n",
    "    class StemCell(Cell): #define stem cell class\n",
    "        def __init__(self):\n",
    "            Cell.__init__(self)\n",
    "            self.pmax = float('inf') #infinite proliferation potential\n",
    "            self.alpha = 0 #immortal\n",
    "            #self.ps = ps #probability of symmetric division\n",
    "            self.proliferation_rate = proliferationCSC\n",
    "\n",
    "\n",
    "    #create lattice array\n",
    "    N = max_cells # starting demensions of lattice\n",
    "    lattice = np.empty( (N,N), dtype=object)\n",
    "\n",
    "    #parameters\n",
    "    proliferation_rate_CC = proliferationCC #proliferation rate of clonal cell\n",
    "    proliferation_rate_CSC = proliferationCSC #proliferation rate of cancer stem cell\n",
    "    pmax_CC = pmax\n",
    "\n",
    "    alpha_CC = mortality\n",
    "    cur_cellnum = 1\n",
    "    cur_mutnum = 1\n",
    "\n",
    "    mutation_rate = mutation\n",
    "    time = 0\n",
    "    \n",
    "    #initialize list to keep track of cells (all time and alive)\n",
    "    alive_cells = []\n",
    "    cells = []\n",
    "    \n",
    "    \n",
    "    nucleotides = [\"A\", \"G\", \"C\", \"T\"]\n",
    "    \n",
    "    #create data frame for location time stamps\n",
    "    #df_empty = pd.DataFrame({'A' : []})\n",
    "    all_locations_df = pd.DataFrame(columns = ['index', 'locx', 'locy', 't'])\n",
    "\n",
    "\n",
    "#to place founder cell in center of lattice\n",
    "    i = int(N/2)\n",
    "    modi = [0,0,1,1]\n",
    "    modj = [0,1,0,1]\n",
    "\n",
    "    for x in range(founder_cells):\n",
    "        Fcell = StemCell() #initiate founder cells, non-stem cells\n",
    "        Fcell.proliferation_rate = proliferation_rate_CC\n",
    "        Fcell.locx = i+modi[x] #x location in lattice\n",
    "        Fcell.locy = i+modj[x] #y location in lattice\n",
    "        Fcell.alpha = mortality\n",
    "\n",
    "        Fcell.birthdate = 0\n",
    "        Fcell.cellnum = cur_cellnum  #unique cell number to identify\n",
    "        cur_cellnum += 1\n",
    "        Fcell.mutation_rate = mutation_rate #probability of mutation per cell division\n",
    "\n",
    "\n",
    "        cells.append(Fcell)\n",
    "        alive_cells.append(Fcell)\n",
    "        lattice[Fcell.locx,Fcell.locy] = Fcell\n",
    "\n",
    "\n",
    "    #time parameters\n",
    "    dt = deltat # time is equilavent to 1/24 of a day or 1 hour\n",
    "#     stop_time =stime#stop simulation after this many days\n",
    "\n",
    "    \n",
    "    #while time < stop_time:\n",
    "    while len(alive_cells) < max_cells:\n",
    "        \n",
    "        if len(alive_cells) < 1:\n",
    "            print('no cells alive')\n",
    "            #start overr simulation\n",
    "            cells = []\n",
    "            alive_cells = []\n",
    "            time = 0\n",
    "            cur_cellnum = 1\n",
    "            cur_mutnum = 1\n",
    "\n",
    "            lattice = np.empty( (N,N), dtype=object)\n",
    "            all_locations_df = pd.DataFrame(columns = ['index', 'locx', 'locy', 't'])\n",
    "            \n",
    "            for x in range(founder_cells):\n",
    "                Fcell = StemCell() #initiate founder cells\n",
    "                Fcell.proliferation_rate = proliferation_rate_CC\n",
    "                Fcell.locx = i+modi[x] #x location in lattice\n",
    "                Fcell.locy = i+modj[x] #y location in lattice\n",
    "                Fcell.alpha = mortality\n",
    "               # Fcell.ps = 0\n",
    "                Fcell.birthdate = 0\n",
    "                Fcell.cellnum = cur_cellnum  #unique cell number to identify\n",
    "                cur_cellnum += 1\n",
    "                Fcell.mutation_rate = mutation_rate #probability of mutation per cell division per time unit\n",
    "\n",
    "\n",
    "                cells.append(Fcell)\n",
    "                alive_cells.append(Fcell)\n",
    "                lattice[Fcell.locx,Fcell.locy] = Fcell\n",
    "            \n",
    "            \n",
    "        time += dt\n",
    "        cell_stack = random.sample(alive_cells,len(alive_cells)) #random order of cells\n",
    "\n",
    "                \n",
    "        time += dt\n",
    "        cell_stack = random.sample(alive_cells, len(alive_cells)) #random order of cells\n",
    "        \n",
    "        for cell in cell_stack:\n",
    "            \n",
    "            #print(cell.cellnum)\n",
    "            #print(cell in alive_cells)\n",
    "            alive = True\n",
    "            r = random.uniform(0, 1) #decision for prolferation\n",
    "            r2 = random.uniform(0, 1) #decision for cell death\n",
    "            p_d = cell.proliferation_rate * dt #probability of proliferation in time dt\n",
    "            \n",
    "\n",
    "            free_cells = adjacent_cell(lattice,cell.locx,cell.locy)\n",
    "            \n",
    "\n",
    "                \n",
    "            p_die = cell.alpha * dt #probability of cell death in time dt\n",
    "            \n",
    "            if r < p_die:\n",
    "                alive = False\n",
    "                lattice[cell.locx,cell.locy] = None\n",
    "                alive_cells.remove(cell)\n",
    "\n",
    "                cell.deathdate = time\n",
    "            \n",
    "                \n",
    "            elif r2 < p_d: # Does cell attempt to divide? \n",
    "\n",
    "                #print(cell in alive_cells)\n",
    "                cell.pmax -= 1\n",
    "                r = random.uniform(0,1)\n",
    "                new_cell = copy.deepcopy(cell)\n",
    "                new_cell_2 = copy.deepcopy(cell)\n",
    "                        \n",
    "                new_cell_2.locx = cell.locx #x location in lattice\n",
    "                new_cell_2.locy = cell.locy #y location in lattice\n",
    "\n",
    "                #alive_cells.remove(cell)\n",
    "                if len(free_cells) > 0:\n",
    "                    \n",
    "                    newloc = random.choice(free_cells)\n",
    "                    new_cell.locx = newloc[0] #x location in lattice\n",
    "                    new_cell.locy = newloc[1] #y location in lattice\n",
    "                    lattice[new_cell.locx, new_cell.locy] = new_cell\n",
    "                    \n",
    "                else:\n",
    "                    \n",
    "                    #choose direction to push (x,y vector)\n",
    "                    \n",
    "                \n",
    "                    pushx = random.choice([-1,0,1])\n",
    "                    pushy = random.choice([-1,0,1])\n",
    "                    \n",
    "                    while pushx == 0 and pushy == 0: #redraw if no shift\n",
    "                        \n",
    "                        pushx = random.choice([-1,0,1])\n",
    "                        pushy = random.choice([-1,0,1])\n",
    "                    \n",
    "                    assert pushx != 0 or pushy != 0, \"Push direction is 0,0\"\n",
    "                    \n",
    "                    new_cell.locx = cell.locx + pushx #x location in lattice\n",
    "                    new_cell.locy = cell.locy + pushy #y location in lattice\n",
    "                    \n",
    "                    lattice[new_cell.locx, new_cell.locy] = new_cell\n",
    "                    \n",
    "                    #then push all other alive cells\n",
    "                    curr_overlap = [new_cell.locx, new_cell.locy]\n",
    "                    \n",
    "                    curr_occupying_cell = new_cell\n",
    "                    \n",
    "                    \n",
    "                    while curr_overlap:\n",
    "                        \n",
    "                        \n",
    "                        \n",
    "                        overlapping_cell = next((x for x in alive_cells if ([x.locx,x.locy] == curr_overlap and x != curr_occupying_cell)), None)\n",
    "                        \n",
    "\n",
    "                        \n",
    "                        if overlapping_cell:\n",
    "                            \n",
    "\n",
    "                        \n",
    "                            curr_free_cells = adjacent_cell(lattice, overlapping_cell.locx, overlapping_cell.locy)\n",
    "                            \n",
    "                            \n",
    "                            if len(curr_free_cells) > 0:\n",
    "                                \n",
    "\n",
    "                                push_newloc = random.choice(curr_free_cells)\n",
    "                                overlapping_cell.locx = push_newloc[0] #x location in lattice\n",
    "                                overlapping_cell.locy = push_newloc[1] #y location in lattice\n",
    "                                curr_overlap = None\n",
    "                                curr_occupying_cell = overlapping_cell\n",
    "\n",
    "                                \n",
    "                            else:    #if there is no empty lattice spot then keep pushing in same direction\n",
    "                                \n",
    "\n",
    "                                overlapping_cell.locx = overlapping_cell.locx + pushx\n",
    "                                overlapping_cell.locy = overlapping_cell.locy + pushy\n",
    "                                curr_overlap = [overlapping_cell.locx, overlapping_cell.locy]\n",
    "                                \n",
    "\n",
    "                                curr_occupying_cell = overlapping_cell\n",
    "                            \n",
    "                            #update lattice and occupying cell\n",
    "                            lattice[overlapping_cell.locx, overlapping_cell.locy] = overlapping_cell\n",
    "                        \n",
    "                        else:\n",
    "                            \n",
    "                            curr_overlap = None\n",
    "                        \n",
    "                new_cell.move = False\n",
    "                new_cell_2.move = False\n",
    "                \n",
    "                new_cell.birthdate = time\n",
    "                new_cell_2.birthdate = time\n",
    "                new_cell.cellnum = cur_cellnum #unique cell number to identify\n",
    "  \n",
    "                \n",
    "               \n",
    "                cur_cellnum += 1\n",
    "                \n",
    "                new_cell_2.cellnum = cur_cellnum\n",
    "                new_cell.parent_index = cell.cellnum\n",
    "                new_cell_2.parent_index = cell.cellnum\n",
    "    \n",
    "                    \n",
    "                cur_cellnum += 1\n",
    "                lattice[cell.locx,cell.locy] = new_cell_2\n",
    "                        \n",
    "\n",
    "                new_cell.cells_free = len(adjacent_cell(lattice,new_cell.locx,new_cell.locy))\n",
    "                new_cell_2.cells_free = len(adjacent_cell(lattice,new_cell_2.locx,new_cell_2.locy))\n",
    "                \n",
    "                r1 = random.uniform(0,1)\n",
    "                if r1 < cell.mutation_rate: #does cell gain a mutation?\n",
    "                            \n",
    "                    if JC: #if Jukes-Cantor model\n",
    "                                \n",
    "                        original_nucleotide = random.choice([\"A\", \"G\", \"C\", \"T\"])\n",
    "                                \n",
    "                        #put original nucleotide in all sequences\n",
    "                        for cell_0 in cells:\n",
    "                            cell_0.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                        #cell 2 has not been added yet\n",
    "                        new_cell_2.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                        #replace mutation in this cells sequence\n",
    "                                \n",
    "                        possible_nucs = [\"A\", \"G\", \"C\", \"T\"]\n",
    "                        possible_nucs.remove(original_nucleotide)\n",
    "                        new_nucleotide = random.choice(possible_nucs)\n",
    "                        new_cell.sequence.append(new_nucleotide)\n",
    "                            \n",
    "                    new_mut = cur_mutnum\n",
    "                    cur_mutnum +=1\n",
    "                    new_cell.mutations.append(new_mut)\n",
    "                        \n",
    "                #does second cell get a mutation\n",
    "                r2 = random.uniform(0,1)\n",
    "                if r2 < cell.mutation_rate: #does cell gain a mutation?\n",
    "                            \n",
    "                    if JC: #if Jukes-Cantor model\n",
    "                                \n",
    "                        original_nucleotide = random.choice([\"A\", \"G\", \"C\", \"T\"])\n",
    "                                \n",
    "                        #put original nucleotide in all sequences\n",
    "                        for cell_0 in cells:\n",
    "                            cell_0.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                        new_cell.sequence.append(original_nucleotide)\n",
    "                                \n",
    "                        possible_nucs = [\"A\", \"G\", \"C\", \"T\"]\n",
    "                        possible_nucs.remove(original_nucleotide)\n",
    "                        new_nucleotide = random.choice(possible_nucs)\n",
    "                        new_cell_2.sequence.append(new_nucleotide)\n",
    "\n",
    "                    new_mut = cur_mutnum\n",
    "                    cur_mutnum +=1\n",
    "                    new_cell_2.mutations.append(new_mut)\n",
    "\n",
    "\n",
    "                alive_cells.append(new_cell)\n",
    "                alive_cells.append(new_cell_2)\n",
    "                cells.append(new_cell)\n",
    "                cells.append(new_cell_2)\n",
    "\n",
    "                alive_cells.remove(cell)\n",
    "\n",
    "                cell.deathdate = time\n",
    "    \n",
    "        #snapshot of tumor locations\n",
    "        location_df = pd.DataFrame([x.get_loc_dict() for x in alive_cells])\n",
    "        location_df['t'] = time\n",
    "        #all_locations_df = all_locations_df.append(location_df, ignore_index = False)\n",
    "        all_locations_df = pd.concat([all_locations_df, location_df], ignore_index = True)\n",
    "        \n",
    "    for cell in alive_cells:\n",
    "\n",
    "        cell.deathdate = time + dt #need to add one time step to differentiate cells that died in last gen\n",
    "\n",
    "\n",
    "    return cells, alive_cells, all_locations_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Boundary-driven growth simulation study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-d4ec45eab888>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     26\u001b[0m         \u001b[0;31m#run simulated\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m         cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n\u001b[0;32m---> 28\u001b[0;31m         proliferationCC = 1, CSC = False, pmax = 10,mutation = 1,mortality = dr, JC = True)\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0;31m#convert to dataframe\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-2-15e763d6ce27>\u001b[0m in \u001b[0;36mcancer_sim\u001b[0;34m(founder_cells, deltat, CSC, max_cells, proliferationCSC, proliferationCC, motility, pmax, mortality, mutation, selection, ps, free_diffusion, JC, s, driverMutCell)\u001b[0m\n\u001b[1;32m    208\u001b[0m                         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muniform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    209\u001b[0m                         \u001b[0mnew_cell\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 210\u001b[0;31m                         \u001b[0mnew_cell_2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    211\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    212\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mr\u001b[0m \u001b[0;34m<=\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mps\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;31m#does cell divide asymmetrically?\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    178\u001b[0m                     \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    179\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 180\u001b[0;31m                     \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_reconstruct\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0mrv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    181\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    182\u001b[0m     \u001b[0;31m# If is its own copy, don't memoize.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36m_reconstruct\u001b[0;34m(x, memo, func, args, state, listiter, dictiter, deepcopy)\u001b[0m\n\u001b[1;32m    278\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mstate\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdeep\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 280\u001b[0;31m             \u001b[0mstate\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstate\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    281\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'__setstate__'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    282\u001b[0m             \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__setstate__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstate\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    148\u001b[0m     \u001b[0mcopier\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_deepcopy_dispatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcopier\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 150\u001b[0;31m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    151\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    152\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36m_deepcopy_dict\u001b[0;34m(x, memo, deepcopy)\u001b[0m\n\u001b[1;32m    238\u001b[0m     \u001b[0mmemo\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 240\u001b[0;31m         \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    241\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdict\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_deepcopy_dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    148\u001b[0m     \u001b[0mcopier\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_deepcopy_dispatch\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    149\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcopier\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 150\u001b[0;31m         \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcopier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    151\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    152\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36m_deepcopy_list\u001b[0;34m(x, memo, deepcopy)\u001b[0m\n\u001b[1;32m    213\u001b[0m     \u001b[0mappend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    214\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0ma\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 215\u001b[0;31m         \u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    216\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    217\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_deepcopy_list\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/copy.py\u001b[0m in \u001b[0;36mdeepcopy\u001b[0;34m(x, memo, _nil)\u001b[0m\n\u001b[1;32m    130\u001b[0m \u001b[0;32mdel\u001b[0m \u001b[0md\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    131\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 132\u001b[0;31m \u001b[0;32mdef\u001b[0m \u001b[0mdeepcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmemo\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_nil\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    133\u001b[0m     \"\"\"Deep copy operation on arbitrary Python objects.\n\u001b[1;32m    134\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "##simulations for x-y plot of estimated diversification rates versus true\n",
    "##this is modified by death rate, keeping proliferation rate constant\n",
    "##\n",
    "random.seed(822156)\n",
    "n_cells = [1000]\n",
    "dr_list = np.arange(0, 0.41, 0.01).tolist()\n",
    "for N in n_cells:\n",
    "    for dr in dr_list:\n",
    "    \n",
    "        #run simulated\n",
    "        cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "        proliferationCC = 1, CSC = False, pmax = 10,mutation = 1,mortality = dr, JC = True)\n",
    "    \n",
    "        #convert to dataframe\n",
    "        CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "        file_name = f\"cells_death_rate_validation_pop_{N}_dr_%0.2f.csv\" %(dr)\n",
    "        CC_df.to_csv(file_name)\n",
    "        \n",
    "random.seed(3718271)\n",
    "n_cells = [1000]\n",
    "dr_list = np.arange(0.41, 0.87, 0.01).tolist()\n",
    "for N in n_cells:\n",
    "    for dr in dr_list:\n",
    "    \n",
    "        #run simulated\n",
    "        cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "        proliferationCC = 1, CSC = False, pmax = 10,mutation = 1,mortality = dr, JC = True)\n",
    "    \n",
    "        #convert to dataframe\n",
    "        CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "        file_name = f\"..cells_death_rate_validation_pop_{N}_dr_%0.2f.csv\" %(dr)\n",
    "        CC_df.to_csv(file_name)\n",
    "        \n",
    "##multiple replicates of same death rates\n",
    "\n",
    "random.seed(31221)\n",
    "n_cells = [1000]\n",
    "dr_list = np.arange(0.05, 0.66, 0.05).tolist()\n",
    "i_list = np.arange(0,10,1).tolist()\n",
    "for N in n_cells:\n",
    "    for dr in dr_list:\n",
    "        for i in i_list:\n",
    "    \n",
    "            #run simulated\n",
    "            cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "            proliferationCC = 1, CSC = False, pmax = 10,mutation = 1,mortality = dr, JC = True)\n",
    "    \n",
    "            #convert to dataframe\n",
    "            CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "            file_name = f\"cells_death_rate_validation_pop_{N}_dr_%0.2f_i_{i}.csv\" %(dr)\n",
    "            CC_df.to_csv(file_name)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'random' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-0f9e92025a08>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m31221\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mn_cells\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m50000\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#run simulated\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'random' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-b5f4a2064162>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m     \u001b[0mfile_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"cells_death_rate_validation_{N}_i_{itr}_dr_%0.2f.csv\"\u001b[0m \u001b[0;34m%\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m     \u001b[0mCC_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_name\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, tupleize_cols, date_format, doublequote, escapechar, decimal)\u001b[0m\n\u001b[1;32m   3018\u001b[0m                                  \u001b[0mdoublequote\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdoublequote\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3019\u001b[0m                                  escapechar=escapechar, decimal=decimal)\n\u001b[0;32m-> 3020\u001b[0;31m         \u001b[0mformatter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3021\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3022\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpath_or_buf\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    170\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwriter\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mUnicodeWriter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mwriter_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 172\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    173\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    174\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    286\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    287\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 288\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/anaconda3/lib/python3.7/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    313\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    314\u001b[0m         libwriters.write_csv_rows(self.data, ix, self.nlevels,\n\u001b[0;32m--> 315\u001b[0;31m                                   self.cols, self.writer)\n\u001b[0m",
      "\u001b[0;32mpandas/_libs/writers.pyx\u001b[0m in \u001b[0;36mpandas._libs.writers.write_csv_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#iterations for branching and clock rate signal analysis\n",
    "random.seed(2121)\n",
    "\n",
    "N=1000\n",
    "\n",
    "dr = 0.1\n",
    "i_list = np.arange(0, 100, 1).tolist()\n",
    "\n",
    "for itr in i_list:\n",
    "    #run simulated\n",
    "    cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "    proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n",
    "    \n",
    "    #convert to dataframe\n",
    "    CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "    file_name = f\"cells_death_rate_validation_{N}_i_{itr}_dr_%0.2f.csv\" %(dr)\n",
    "    CC_df.to_csv(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unrestricted growth simulation study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:358: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells alive\n",
      "no cells alive\n",
      "0.0\n",
      "0.01\n",
      "0.02\n",
      "0.03\n",
      "0.04\n",
      "0.05\n",
      "0.06\n",
      "0.07\n",
      "0.08\n",
      "0.09\n",
      "0.1\n",
      "0.11\n",
      "no cells alive\n",
      "0.12\n",
      "0.13\n",
      "0.14\n",
      "no cells alive\n",
      "0.15\n",
      "0.16\n",
      "0.17\n",
      "0.18\n",
      "no cells alive\n",
      "0.19\n",
      "0.2\n",
      "0.21\n",
      "0.22\n",
      "0.23\n",
      "no cells alive\n",
      "0.24\n",
      "no cells alive\n",
      "0.25\n",
      "0.26\n",
      "no cells alive\n",
      "0.27\n",
      "no cells alive\n",
      "0.28\n",
      "no cells alive\n",
      "0.29\n",
      "0.3\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "0.31\n",
      "no cells alive\n",
      "0.32\n",
      "0.33\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "0.34\n",
      "0.35000000000000003\n",
      "0.36\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "0.37\n",
      "no cells alive\n",
      "0.38\n",
      "no cells alive\n",
      "0.39\n",
      "0.4\n",
      "no cells alive\n",
      "no cells alive\n",
      "0.41000000000000003\n",
      "0.42\n",
      "no cells alive\n",
      "0.43\n",
      "no cells alive\n",
      "0.44\n",
      "0.45\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "0.46\n",
      "0.47000000000000003\n",
      "no cells alive\n",
      "0.48\n",
      "no cells alive\n",
      "0.49\n",
      "0.5\n",
      "no cells alive\n",
      "0.51\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-5ac60844b611>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     32\u001b[0m     \u001b[0;31m#run simulated\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m     cells_CC, alive_cells_CC, all_locations_df_CC = pushing_cancer_sim2(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n\u001b[0;32m---> 34\u001b[0;31m     proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n\u001b[0m\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0;31m#save locations\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-45537087db0a>\u001b[0m in \u001b[0;36mpushing_cancer_sim2\u001b[0;34m(founder_cells, deltat, CSC, max_cells, proliferationCSC, proliferationCC, motility, pmax, mortality, mutation, selection, JC)\u001b[0m\n\u001b[1;32m    199\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    200\u001b[0m                     \u001b[0mnewloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchoice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfree_cells\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 201\u001b[0;31m                     \u001b[0mnew_cell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnewloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m#x location in lattice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    202\u001b[0m                     \u001b[0mnew_cell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnewloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;31m#y location in lattice\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m                     \u001b[0mlattice\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnew_cell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnew_cell\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlocy\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnew_cell\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#run multiple pushing simulations at dr = 0.1\n",
    "random.seed(9918)\n",
    "\n",
    "N=1000\n",
    "\n",
    "dr = 0.1\n",
    "i_list = np.arange(0, 10, 1).tolist()\n",
    "\n",
    "for itr in i_list:\n",
    "    #run simulated\n",
    "    cells_CC, alive_cells_CC, all_locations_df_CC = pushing_cancer_sim2(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "    proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n",
    "    \n",
    "    #save locations\n",
    "    locs_file_name = f\"cells_pushing_pop2_{N}_dr_%0.2f_i_{itr}_locs.csv\" %(dr)\n",
    "    all_locations_df_CC.to_csv(locs_file_name)\n",
    "    #convert to dataframe\n",
    "    CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "    file_name = f\"cells_pushing_pop2_{N}_dr_%0.2f_i_{itr}.csv\" %(dr)\n",
    "    CC_df.to_csv(file_name)\n",
    "\n",
    "#Run one iteration accross range of death rates\n",
    "random.seed(8161)\n",
    "\n",
    "N=1000\n",
    "\n",
    "dr_list = np.arange(0, 0.88, 0.01).tolist()\n",
    "\n",
    "for dr in dr_list:\n",
    "    print(dr)\n",
    "    #run simulated\n",
    "    cells_CC, alive_cells_CC, all_locations_df_CC = pushing_cancer_sim2(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "    proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n",
    "    \n",
    "    #save locations\n",
    "    locs_file_name = f\"cells_pushing_pop_{N}_dr_%0.2f_locs.csv\" %(dr)\n",
    "    all_locations_df_CC.to_csv(locs_file_name)\n",
    "    #convert to dataframe\n",
    "\n",
    "    #convert to dataframe\n",
    "    CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "    file_name = f\"cells_pushing_pop_{N}_dr_%0.2f.csv\" %(dr)\n",
    "    CC_df.to_csv(file_name)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:358: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n"
     ]
    }
   ],
   "source": [
    "#run all pushing validations\n",
    "#interations\n",
    "#intentionally same seed as above\n",
    "random.seed(1211)\n",
    "\n",
    "N=1000\n",
    "\n",
    "dr_list = np.arange(0.05, 0.66, 0.05).tolist()\n",
    "i_list = np.arange(0,10,1).tolist()\n",
    "\n",
    "for dr in dr_list:\n",
    "    for i in i_list:\n",
    "        #run simulated\n",
    "        cells_CC, alive_cells_CC, all_locations_df_CC = pushing_cancer_sim2(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "        proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n",
    "        \n",
    "        \n",
    "        #save locations\n",
    "        locs_file_name = f\"cells_pushing_pop_{N}_dr_%0.2f_i_{i}_locs.csv\" %(dr)\n",
    "        all_locations_df_CC.to_csv(locs_file_name)\n",
    "        #convert to dataframe\n",
    "        CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "        file_name = f\"cells_pushing_pop_{N}_dr_%0.2f_i_{i}.csv\" %(dr)\n",
    "        CC_df.to_csv(file_name)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:358: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n",
      "no cells alive\n"
     ]
    }
   ],
   "source": [
    "#run all pushing validations at dr = 0.1\n",
    "random.seed(2121)\n",
    "\n",
    "N=1000\n",
    "\n",
    "dr = 0.1\n",
    "i_list = np.arange(0, 100, 1).tolist()\n",
    "\n",
    "for itr in i_list:\n",
    "    #run simulated\n",
    "    cells_CC, alive_cells_CC, lattice_CC = pushing_cancer_sim(founder_cells = 1,max_cells = N, proliferationCSC = 1,\n",
    "    proliferationCC = 1, CSC = False, pmax = 10,mutation = 1, mortality = dr, JC = True)\n",
    "    \n",
    "    #convert to dataframe\n",
    "    CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "    file_name = f\"cells_pushing_pop_{N}_i_{itr}_dr_%0.2f.csv\" %(dr)\n",
    "    CC_df.to_csv(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mathematical Oncology Coverart Tumor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(31221)\n",
    "n_cells = 15000\n",
    "dr = 0.1\n",
    "    \n",
    "#run simulated\n",
    "cells_CC, alive_cells_CC, lattice_CC = cancer_sim(founder_cells = 1,max_cells = n_cells, proliferationCSC = 1,\n",
    "proliferationCC = 1, CSC = False, pmax = 10,mutation = 0.1,mortality = dr, JC = True)\n",
    "    \n",
    "#convert to dataframe\n",
    "CC_df = pd.DataFrame([x.as_dict() for x in cells_CC])\n",
    "    \n",
    "file_name = f\"cells_death_rate_validation_pop_{n_cells}_dr_%0.2f.csv\" %(dr)\n",
    "CC_df.to_csv(file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
